Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIScriptingCocoa.mm

```diff
@@ -0,0 +1,255 @@
+        funcKey: NSObject.class,
+        functionKey : NSObject.class,
+        targetKey: NSDictionary.class,
+        worldKey: NSString.class,
+    };
+
+    if (!validateDictionary(script, nil, requiredKeys, keyTypes, outExceptionString))
+        return false;
+
+    if (!validateTarget(script[targetKey], outExceptionString))
+        return false;
+
+    if (script[functionKey] && script[funcKey]) {
+        *outExceptionString = toErrorString(nil, nil, @"Both 'func' and 'function' cannot be passed. Please use 'func'.");
+        return false;
+    }
+
+    if (script[argumentsKey] && script[argsKey]) {
+        *outExceptionString = toErrorString(nil, nil, @"Both 'args' and 'arguments' cannot be passed. Please use 'args'.");
+        return false;
+    }
+
+    bool functionWasPassed = script[functionKey] || script[funcKey];
+    if ((script[filesKey] && functionWasPassed)
+        || (!functionWasPassed && !script[filesKey])) {
+        *outExceptionString = toErrorString(nil, nil, @"Either 'func' or 'files' must be passed, but not both at the same time.");
+        return false;
+    }
+
+    BOOL scriptContainsArguments = !!(script[argsKey] || script[argumentsKey]);
+    if (scriptContainsArguments && !functionWasPassed) {
+        *outExceptionString = toErrorString(nil, nil, @"The 'args' property must be passed along with the 'func' property.");
+        return false;
+    }
+
+    if (script[filesKey]) {
+        NSArray *files = script[filesKey];
+        if (!files.count) {
+            *outExceptionString = toErrorString(nil, filesKey, @"at least one file must be passed.");
+            return false;
+        }
+    }
+
+    if (NSString *world = script[worldKey]) {
+        if (![world isEqualToString:isolatedWorld] && ![world isEqualToString:mainWorld]) {
+            *outExceptionString = toErrorString(nil, worldKey, @"it must either have the value of 'ISOLATED' or 'MAIN'.");
+            return false;
+        }
+    }
+
+    return true;
+}
+
+bool WebExtensionAPIScripting::validateTarget(NSDictionary *targetInfo, NSString **outExceptionString)
+{
+    static NSArray<NSString *> *requiredKeys = @[
+        tabIDKey,
+    ];
+    static NSDictionary<NSString *, id> *keyTypes = @{
+        frameIDsKey: @[ NSNumber.class ],
+        tabIDKey: NSNumber.class,
+        allFramesKey: @YES.class,
+    };
+
+    if (!validateDictionary(targetInfo, nil, requiredKeys, keyTypes, outExceptionString))
+        return false;
+
+    if (targetInfo[allFramesKey] && targetInfo[frameIDsKey]) {
+        *outExceptionString = toErrorString(nil, nil, @"The 'allFrames' and 'frameIds' properties cannot both be specified.");
+        return false;
+    }
+
+    NSArray *frameIDs = targetInfo[frameIDsKey];
+    for (NSNumber *frameID in frameIDs) {
+        if (frameID.longLongValue < mainFrameID) {
+            *outExceptionString = toErrorString(nil, frameIDsKey, @"at least one frame id is not a positive number.");
```

Comment:
Can drop the trailing period, the helper function adds it if needed.

Reply:
I would just say `@"it is not a frame identifier"` to cover all the bases. This matches tab and window ID errors.

Reply:
Better yet, since this is an array, I have been including the problem value.

```
toErrorString(nil, frameIDsKey, @"'%@' is not a frame identifier", frameID)
```